Idempotent publishing

jdavid opened the pull request:
Idempotent publishing

jdavid commented:
RSL1k4 and RSL1k5 are failing, for instance https://travis-ci.org/ably/ably-python/jobs/415827004#L696

So the 3 messages are published even when they have the same id, any idea why?

mattheworiordan commented on test/ably/restchannelpublish_test.py:
```diff
@@ -425,3 +432,19 @@ def test_idempotent_rest_publishing(self):
+        channel = self.get_channel()
+
+        data = {
+            'name': 'name',
```
Given this change is about idempotent publishing, are not checking that `id` carries through as well?

mattheworiordan commented on test/ably/restchannelpublish_test.py:
```diff
@@ -508,3 +508,28 @@ def side_effect(self, *args, **kwargs):
         assert failures == 2
         assert len(channel.history().items) == 1
+
+    # RSL1k5
```
I don't think this test is correct.  We are not talking about error conditions in the spec, we are requiring that the same message is published many times successfully, and then we confirm only one message is stored.

jdavid commented:
The logs when running pytest with log level debug:
https://travis-ci.org/ably/ably-python/jobs/417173671#L796

Need more detail maybe?

mattheworiordan commented:
Yes, we do need more detail.

The only way we can help you debug the issue is if we can see a lot of HTTP requests and responses.  Can you include that in the debug log level?

jdavid commented:
https://travis-ci.org/ably/ably-python/jobs/418271334#L780

jdavid commented:
ping

mattheworiordan commented:
@jdavid but you've not addressed all my feedback?

jdavid commented:
Yes, the first comment is addressed in commit 99bf9b0c (next time I'll pay attention to make the change so the review comment becomes *outdated* automatically)

Then added the request/response details to the log, see https://travis-ci.org/ably/ably-python/jobs/418271334#L780

```
< POST /channels/tdTHrpjtEE/publish HTTP/1.1
< Host: sandbox-rest.ably.io:443
< Connection: keep-alive
< Accept-Encoding: gzip, deflate
< Accept: application/json
< User-Agent: python-requests/2.19.1
< X-Ably-Lib: python-1.0.1
< X-Ably-Version: 1.0
< Content-Type: application/json
< Authorization: Basic X3RtcF9ST2l1cEEuV29HMnNROklwTkVIV25HdktDV0gzcGM=
< Content-Length: 71
< 
< {"timestamp":1534777245486,"data":"data1","name":"name1","id":"foobar"}
> HTTP/1.1 201 Created
> Access-Control-Allow-Credentials: true
> Access-Control-Allow-Origin: *
> Access-Control-Expose-Headers: Link,Transfer-Encoding,Content-Length,X-Ably-ErrorCode,X-Ably-ErrorMessage,X-Ably-ServerId,Server
> Content-Length: 58
> Content-Type: application/json
> Date: Mon, 20 Aug 2018 15:00:45 GMT
> Vary: Origin
> X-Ably-Serverid: frontend.7d44.2.us-east-1-A.i-0a5a6ad4bd3f037bc
> 
{
	"channel": "tdTHrpjtEE",
	"messageId": "K4Y_NQjK6f:0"
}
```

It's sending `id=foobar` and the server replies with `messageId=K4Y_NQjK6f:0`

Guess that's the problem. Should it send `messageId` instead of `id`?

jdavid commented:
Oh yes, it was copy/paste from RSL1k4, fixed now.

Changed to use the `/messages` endpoint, but now there're a few test failures that worked fine with `/publish`, server sending a 50000 error, https://travis-ci.org/ably/ably-python/jobs/419670479#L628

mattheworiordan commented:
> but now there're a few test failures that worked fine with /publish

Those appear to be connection refused failures, which I assume are probably related to some other network issue.  Can you see if those errors happen again or not?

paddybyers commented:
> In terms of what you are sending, that looks OK. I fear that sandbox has perhaps now got a version running that no longer supports idempotency :( We do have dev feature branch environments to avoid these types of issues. @paddybyers can you confirm what's going on here and if there is a different env @jdavid can use?

It's not running on sandbox now. I can deploy that to `dev` until we get it merged.

jdavid commented:
Except https://travis-ci.org/ably/ably-python/jobs/419670480 which has had an unrelated and complete failure, the other 7 test builds have the same 3 failures in `test/ably/restauth_test.py`

Also I've run these tests several times locally, switching back and forward from `/publish` to `/messages`. Always working with `/publish`, always failing with `/messages` ; will retry tomorrow.

paddybyers commented:
> is there anything different about those REST routes?

I can't see any difference, no, assuming that the method is `POST`. `GET` will be different; `GET /messages` is history, not publish.

paddybyers commented:
> I can deploy that to dev until we get it merged.

This is now deployed. You can access the `dev` environment with a hostname of `dev-rest.ably.io`, the same as what you should get with client option `environment: 'dev'`.

jdavid commented:
Yes, it's POST.

The I realized that we were using the old endpoint for the history as well, so changed it from `GET /channels/<channel id>/history` to `GET /channels/<channel id>/messages`, and now there're 2 more tests failing in https://travis-ci.org/ably/ably-python/jobs/420010317#L593 , with 50000 errors as well.

jdavid commented:
Using `dev` instead of `sandbox`fixes the idempotent tests.

Do I change back to use the old endpoints to make the tests pass?

@paddybyers There's still the @mattheworiordan question about `timestamp`:

> Secondly, I am not sure why you're sending a timestamp, that is never used. Unless I am mistaken, you should not be sending that. Can you confirm @paddybyers?

paddybyers commented:
> Do I change back to use the old endpoints to make the tests pass?

I'm not sure what you mean by "old endpoints". I think you should keep it as `dev` in this branch until we have a permanent arrangement. It sounds like everything's passing there?

> Secondly, I am not sure why you're sending a timestamp, that is never used. Unless I am mistaken, you should not be sending that. Can you confirm @paddybyers?

Correct. I've checked the spec and it's not really clear enough. The library doesn't populate `timestamp` when publishing messages.

jdavid commented:
Okay, the timestamp is done, it's not being sent anymore.

By the endpoints I mean, if the library uses:

    POST /channels/<channel>/publish
    GET /channels/<channel>/history

Then the tests pass. But if the library uses:

    POST /channels/<channel>/messages
    GET /channels/<channel>/messages

Then there're 5 test failures.

SimonWoolf commented:
> 3 failures in test/ably/restauth_test.py ... Always working with /publish, always failing with /messages

Was curious about this as realtime passes literally the same callback for both endpoints (for POSTs at least). So had a quick look, and... well, I'm not a python guy, but it seems potentially relevant that those tests [mock out the /publish endpoint](https://github.com/ably/ably-python/blob/develop/test/ably/restauth_test.py#L489-L493).

jdavid commented:
Good catch @SimonWoolf , that was it.

Fixed another couple of bugs in Python 2 for idempotent.

Now the tests are almost green, except for one timeout, and again the old random errors in a couple of test runs.

jdavid commented:
ping

The failing tests https://travis-ci.org/ably/ably-python/jobs/427137745#L593 are unrelated to the idempotent feature (the random failures again). Do we merge into `develop`like we did with push? Or should I try to fix them in the context of this PR?

mattheworiordan commented:
Let's merge and fix as a separate PR please.
